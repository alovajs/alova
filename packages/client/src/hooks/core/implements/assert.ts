import { EnumHookType } from '@/util/helper';
import { createAssert, instanceOf } from '@alova/shared';
import { AlovaGenerics, Method } from 'alova';

export const coreAssert = createAssert('');
export const requestHookAssert = createAssert('useRequest');
export const watcherHookAssert = createAssert('useWatcher');
export const fetcherHookAssert = createAssert('useFetcher');

export const coreHookAssert = (hookType: EnumHookType) =>
  ({
    [EnumHookType.USE_REQUEST]: requestHookAssert,
    [EnumHookType.USE_WATCHER]: watcherHookAssert,
    [EnumHookType.USE_FETCHER]: fetcherHookAssert
  })[hookType];

/**
 * Assert whether it is a method instance
 * @param methodInstance method instance
 */
export const assertMethod = <AG extends AlovaGenerics>(assert: typeof requestHookAssert, methodInstance?: Method<AG>) =>
  assert(instanceOf(methodInstance, Method), 'expected a method instance.');
