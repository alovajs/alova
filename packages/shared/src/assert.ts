import { newInstance } from './function';

/**
 * alova error class
 */
export class AlovaError extends Error {
  constructor(prefix: string, message: string, errorCode?: number) {
    super(message + (errorCode ? `\n\nFor detailed: https://alova.js.org/error#${errorCode}` : ''));
    this.name = `[alova${prefix ? `/${prefix}` : ''}]`;
  }
}

/**
 * Custom assertion function that throws an error when the expression is false
 * When errorCode is passed in, a link to the error document will be provided to guide the user to correct it.
 * @param expression Judgment expression, true or false
 * @param message Assert message
 */
export const createAssert =
  (prefix = '') =>
  (expression: any, message: string, errorCode?: number): asserts expression => {
    if (!expression) {
      throw newInstance(AlovaError, prefix, message, errorCode);
    }
  };
