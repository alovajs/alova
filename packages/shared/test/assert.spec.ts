import { AlovaError, createAssert } from '@/assert';

describe('createAssert', () => {
  test('alova error', () => {
    const error1 = new AlovaError('', 'test message');
    expect(error1.name).toBe('[alova]');
    expect(error1.message).toBe('test message');

    const error2 = new AlovaError('ttt', 'test message', 1001);
    expect(error2.name).toBe('[alova/ttt]');
    expect(error2.message).toBe('test message\n\nFor detailed: https://alova.js.org/error#1001');
  });

  test('should throw an error when expression is false with prefix', () => {
    const assert = createAssert('validation');
    expect(() => assert(false, 'Value is not valid')).toThrowError('Value is not valid');
  });

  test('should not throw an error when expression is true with prefix', () => {
    const assert = createAssert('validation');
    expect(() => assert(true, 'Value is valid')).not.toThrow();
  });

  test('should throw an error with error code and include error details URL', () => {
    const assert = createAssert();
    const errCode = 1001;
    const msg = 'Invalid configuration';
    expect(() => assert(false, msg, errCode)).toThrowError(
      'Invalid configuration\n\nFor detailed: https://alova.js.org/error#1001'
    );
  });

  test('should throw an error without error code', () => {
    const assert = createAssert();
    const msg = 'Something went wrong';
    expect(() => assert(false, msg)).toThrowError('Something went wrong');
  });

  test('should handle null/undefined prefix gracefully', () => {
    const assertWithNull = createAssert(null as any);
    const assertWithUndefined = createAssert(undefined);
    expect(() => assertWithNull(false, 'Test with null prefix')).toThrowError('Test with null prefix');
    expect(() => assertWithUndefined(false, 'Test with undefined prefix')).toThrowError('Test with undefined prefix');
  });
});
