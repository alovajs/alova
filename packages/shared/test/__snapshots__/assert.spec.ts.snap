// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createAssert > should handle null/undefined prefix gracefully 1`] = `[[alova]: Test with null prefix]`;

exports[`createAssert > should handle null/undefined prefix gracefully 2`] = `[[alova]: Test with undefined prefix]`;

exports[`createAssert > should throw an error when expression is false with prefix 1`] = `[[alova/validation]: Value is not valid]`;

exports[`createAssert > should throw an error with error code and include error details URL 1`] = `
[[alova]: Invalid configuration

For detailed: https://alova.js.org/error#1001]
`;

exports[`createAssert > should throw an error without error code 1`] = `[[alova]: Something went wrong]`;

exports[`createAssert should handle null/undefined prefix gracefully 1`] = `"Test with null prefix"`;

exports[`createAssert should handle null/undefined prefix gracefully 2`] = `"Test with undefined prefix"`;

exports[`createAssert should throw an error when expression is false with prefix 1`] = `"Value is not valid"`;

exports[`createAssert should throw an error with error code and include error details URL 1`] = `
"Invalid configuration

For detailed: https://alova.js.org/error#1001"
`;

exports[`createAssert should throw an error without error code 1`] = `"Something went wrong"`;
